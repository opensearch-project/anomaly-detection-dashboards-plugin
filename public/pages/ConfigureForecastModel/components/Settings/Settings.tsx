/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

import React, { Fragment, useEffect, useRef, useState } from 'react';
import { Field, FieldProps, FormikProps } from 'formik';
import {
  EuiSpacer,
  EuiFlexGroup,
  EuiFlexItem,
  EuiBadge,
  EuiCompressedFieldNumber,
  EuiRange,
  EuiText,
  EuiPanel,
} from '@elastic/eui';
import {
  isInvalid,
  getError,
  validatePositiveInteger,
  validateNonNegativeInteger,
  validateHistory,
} from '../../../../utils/utils';
import { FORECASTER_DOCS_LINK, FIELD_MAX_WIDTH, INPUT_SLIDER_WIDTH } from '../../../../utils/constants';
import { FormattedFormRow } from '../../../../components/FormattedFormRow/FormattedFormRow';
import '../../index.scss';
import { toNumberOrEmpty } from '../../utils/helpers';

interface SettingsProps {
  isEditable?: boolean;
}

export const Settings = ({ isEditable = true }: SettingsProps) => {
  return (
    <Fragment>
      {/* FORECASTING INTERVAL */}
      <Field name="interval" validate={validatePositiveInteger}>
        {({ field, form }: FieldProps<number>) => {
          const value = field.value ?? '';
          return (
            <FormattedFormRow
              fullWidth
              title="Forecasting interval"
              hint={['How often the forecast runs to generate next value.']}
              hintLink={FORECASTER_DOCS_LINK}
              isInvalid={isInvalid(field.name, form)}
              error={getError(field.name, form)}
              // FIXME: We are going to lift the 1 hour interval restriction.
              helpText="The interval must be at least one minute."
            >
              <EuiFlexGroup
                gutterSize="none"
                alignItems="center"
                style={{ maxWidth: FIELD_MAX_WIDTH }}
              >
                <EuiFlexItem grow={false}>
                  <div style={{ width: INPUT_SLIDER_WIDTH }}>
                    <EuiCompressedFieldNumber
                      name={field.name}
                      id={field.name}
                      value={value}
                      style={{ width: '100%' }}
                      disabled={!isEditable}
                      onChange={(e) => {
                        form.setFieldValue(field.name, toNumberOrEmpty(e.target.value));
                      }}
                    />
                  </div>
                </EuiFlexItem>
                <EuiFlexItem grow={false}>
                  <EuiBadge color="default" className="unit-badge">
                    minutes
                  </EuiBadge>
                </EuiFlexItem>
              </EuiFlexGroup>
            </FormattedFormRow>
          );
        }}
      </Field>

      <EuiSpacer size="l" />

      {/* WINDOW DELAY */}
      <Field name="windowDelay" validate={validateNonNegativeInteger}>
        {({ field, form }: FieldProps<number>) => {
          const value = field.value ?? '';
          return (
            <FormattedFormRow
              fullWidth
              title="Window delay"
              hint="Specify a window of delay for a forecaster to fetch data, if you need to account for extra processing time."
              hintLink={FORECASTER_DOCS_LINK}
              isInvalid={isInvalid(field.name, form)}
              error={getError(field.name, form)}
            >
              <EuiFlexGroup
                gutterSize="none"
                alignItems="center"
                style={{ maxWidth: FIELD_MAX_WIDTH }}
              >
                {/* grow={false} in both the numeric field and unitprevents horizontal expansion,
                 keeping the numeric field at its natural width and allows the unit badge to sit directly next to it.
                 Otherwise, there is a gap between the numeric field and the unit badge. */}
                <EuiFlexItem grow={false}>
                  <div style={{ width: INPUT_SLIDER_WIDTH }}>
                    <EuiCompressedFieldNumber
                      name={field.name}
                      id={field.name}
                      value={value}
                      style={{ width: '100%' }}
                      disabled={!isEditable}
                      onChange={(e) => {
                        form.setFieldValue(field.name, toNumberOrEmpty(e.target.value));
                      }}
                    />
                  </div>
                </EuiFlexItem>
                <EuiFlexItem grow={false}>
                  <EuiBadge color="default" className="unit-badge">
                    minutes
                  </EuiBadge>
                </EuiFlexItem>
              </EuiFlexGroup>
            </FormattedFormRow>
          );
        }}
      </Field>

      <EuiSpacer size="l" />

      <Field name="horizon" validate={validatePositiveInteger}>
        {({ field, form }: FieldProps<number>) => {
          const [showHorizonDetails, setShowHorizonDetails] = useState(false);

          const interval = (form as FormikProps<any>).values.interval;
          const horizon = field.value ?? 0;
          const value = field.value ?? '';

          // Ref to the SLIDER area; if a click happens outside this ref, we close the slider
          const sliderPanelRef = useRef<HTMLDivElement>(null);

          // UseEffect that only attaches the "outside click" listener when the slider is open
          useEffect(() => {
            if (!showHorizonDetails) return;

            function handleOutsideClick(e: MouseEvent) {
              // If the click is not inside sliderPanelRef => close slider
              if (
                sliderPanelRef.current &&
                !sliderPanelRef.current.contains(e.target as Node)
              ) {
                setShowHorizonDetails(false);
              }
            }

            // Note we use capture = true to catch the event early. Could also omit if it works.
            document.addEventListener('mousedown', handleOutsideClick, true);
            return () => {
              document.removeEventListener('mousedown', handleOutsideClick, true);
            };
          }, [showHorizonDetails]);

          // 2) If user moves slider => update field
          function onSliderChange(newVal: number) {
            form.setFieldValue(field.name, newVal);
          }

          // For help text
          function getHorizonHelpText(h: number, i: number) {
            if (!i || !h) return 'A valid horizon is between 1 and 180.';
            const totalMinutes = h * i;
            const hours = Math.floor(totalMinutes / 60);
            const minutes = totalMinutes % 60;
            return (
              `${h} intervals = ${hours} hour${hours === 1 ? '' : 's'}${minutes ? ` ${minutes} minute${minutes === 1 ? '' : 's'}` : ''
              } if the forecasting interval is ${i} minutes. A valid horizon is between 1 and 180.`
            );
          }

          return (
            <FormattedFormRow
              fullWidth
              title="Horizon"
              hint={['How far the forecast extends into the future.']}
              hintLink={FORECASTER_DOCS_LINK}
              isInvalid={isInvalid(field.name, form)}
              error={getError(field.name, form)}
              helpText={
                !showHorizonDetails
                  ? 'Click the number field for more options.'
                  : getHorizonHelpText(horizon, interval)
              }
            >
              {/* 
          Constrain the entire row for consistency with other fields. 
          The numeric field + badge is outside the slider ref. 
        */}
              <div style={{ maxWidth: FIELD_MAX_WIDTH }}>
                <EuiFlexGroup gutterSize="none" alignItems="center">
                  <EuiFlexItem grow={false}>
                    {/* 3) onMouseDown to open the slider, with e.stopPropagation() */}
                    <div style={{ width: INPUT_SLIDER_WIDTH }}>
                      <EuiCompressedFieldNumber
                        min={1}
                        max={180}
                        id={field.name}
                        value={value}
                        style={{ width: '100%' }}
                        disabled={!isEditable}
                        // Use onMouseDown so we can stopPropagation
                        onMouseDown={(e) => {
                          // Prevent this click from bubbling to outside-click handler
                          e.stopPropagation();
                          setShowHorizonDetails(true);
                        }}
                        onChange={(e) => {
                          form.setFieldValue(field.name, toNumberOrEmpty(e.target.value));
                        }}
                      />
                    </div>
                  </EuiFlexItem>
                  <EuiFlexItem grow={false}>
                    <EuiBadge color="default" className="unit-badge">
                      intervals
                    </EuiBadge>
                  </EuiFlexItem>
                </EuiFlexGroup>

                {/* 
            Conditionally render the slider in a ref-wrapped div 
               so that clicks inside don't close the slider. 
          */}
                {showHorizonDetails && (
                  <>
                    <EuiSpacer size="s" />
                    <div ref={sliderPanelRef}>
                      <EuiPanel hasBorder hasShadow={false} paddingSize="m" style={{ width: INPUT_SLIDER_WIDTH }}>
                        <EuiRange
                          style={{ width: '100%' }}
                          min={1}
                          max={180}
                          step={1}
                          value={String(horizon)}
                          onChange={(e) => onSliderChange(Number(e.currentTarget.value))}
                          showLabels
                          showRange
                          levels={[
                            { min: 1, max: 40, color: 'success' },
                            { min: 40, max: 180, color: 'warning' },
                          ]}
                          aria-label="Horizon slider"
                        />
                        <EuiSpacer size="xs" />
                        <EuiText size="xs" color="subdued">
                          Recommended horizon is up to 40 intervals.
                        </EuiText>
                      </EuiPanel>
                    </div>
                  </>
                )}
              </div>
            </FormattedFormRow>
          );
        }}
      </Field>

      <EuiSpacer size="l" />

      {/* HISTORY */}
      <Field name="history" validate={validateHistory}>
        {({ field, form }: FieldProps<number>) => {
          const interval = (form as FormikProps<any>).values.interval;
          const value = field.value ?? '';

          function getHistoryHelpText(history: number, intv: number) {
            if (!intv || !history) return 'Minimum history: 40 intervals.';
            const totalMinutes = history * intv;
            const hours = Math.floor(totalMinutes / 60);
            const mins = totalMinutes % 60;
            return `${history} intervals = ${hours} hour${
              hours === 1 ? '' : 's'
            }${
              mins ? ` ${mins} minute${mins === 1 ? '' : 's'}` : ''
            } if the forecasting interval is ${intv} minutes. Minimum history: 40 intervals.`;
          }

          return (
            <FormattedFormRow
              fullWidth
              title="History"
              hint={['How far back the model looks for training data.']}
              hintLink={FORECASTER_DOCS_LINK}
              isInvalid={isInvalid(field.name, form)}
              error={getError(field.name, form)}
              helpText={getHistoryHelpText(field.value, interval)}
            >
              {/* gutterSize="none" ensures there's no default spacing between flex items */}
              <EuiFlexGroup
                gutterSize="none"
                alignItems="center"
                style={{ maxWidth: FIELD_MAX_WIDTH }}
              >
                
                <EuiFlexItem grow={false}>
                  <div style={{ width: INPUT_SLIDER_WIDTH }}>
                    <EuiCompressedFieldNumber
                      min={40}
                      max={10000}
                      name={field.name}
                      id={field.name}
                      value={value}
                      disabled={!isEditable}
                    style={{ width: '100%' }}
                    onChange={(e) => {
                      form.setFieldValue(field.name, toNumberOrEmpty(e.target.value));
                    }}
                  />
                  </div>
                </EuiFlexItem>
                <EuiFlexItem grow={false}>
                  <EuiBadge color="default" className="unit-badge">
                    intervals
                  </EuiBadge>
                </EuiFlexItem>
              </EuiFlexGroup>
            </FormattedFormRow>
          );
        }}
      </Field>
    </Fragment>
  );
};
