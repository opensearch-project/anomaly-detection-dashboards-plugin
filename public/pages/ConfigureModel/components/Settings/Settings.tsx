/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

import {
  EuiCompressedFieldNumber,
  EuiFlexGroup,
  EuiFlexItem,
  EuiText,
} from '@elastic/eui';
import { Field, FieldProps, useFormikContext } from 'formik';
import React, { Fragment } from 'react';
import {
  isInvalid,
  getError,
  validatePositiveInteger,
  validateMultipleOf,
  validateEmptyOrPositiveInteger,
  validateEmptyOrNonNegativeInteger,
} from '../../../../utils/utils';
import { AD_DOCS_LINK } from '../../../../utils/constants';
import { FormattedFormRow } from '../../../../components/FormattedFormRow/FormattedFormRow';

export const Settings = () => {
  const formik = useFormikContext();
  const numberFieldWidth = 140;

  // Custom validation function that has access to form values
  const validateFrequency = (value: any) => {
    return validateMultipleOf(value, formik.values.interval);
  };

  return (
    <Fragment>
      <Field name="interval" validate={validatePositiveInteger}>
        {({ field, form }: FieldProps) => {
          // Keep frequency in sync when the interval changes (only when they started equal).
          // This reduces UX friction for users who don't care to tweak frequency while editing interval.
          const handleIntervalChange = (
            event: React.ChangeEvent<HTMLInputElement>
          ) => {
            const raw = event.target.value;
            const nextInterval = raw === '' ? '' : Number(raw);

            // remember the values before we mutate anything
            const prevInterval = field.value;
            const prevFrequency = form.values.frequency;

            // push the new interval into Formik
            form.setFieldValue(field.name, nextInterval);

            // if frequency was just mirroring interval, keep it in lockstep
            const prevIntervalNumber = Number(prevInterval);
            const prevFrequencyNumber = Number(prevFrequency);
            const wasSynced =
              !Number.isNaN(prevIntervalNumber) &&
              prevIntervalNumber === prevFrequencyNumber;

            if (wasSynced) {
              form.setFieldValue('frequency', nextInterval);
            }
          };

          return (
            <EuiFlexGroup>
              <EuiFlexItem style={{ maxWidth: '70%' }}>
                <FormattedFormRow
                  fullWidth
                  title="Interval"
                  hint={[
                    `Interval sets the time window for summarizing and modeling data (e.g., 5 min to 1 hr), where too small creates noise, higher cost, and overreaction to fluctuations, while too large smooths out anomalies and delays detection.`,
                  ]}
                  hintLink={`${AD_DOCS_LINK}`}
                  isInvalid={isInvalid(field.name, form)}
                  error={getError(field.name, form)}
                >
                  <EuiFlexGroup gutterSize="s" alignItems="center">
                    <EuiFlexItem grow={false}>
                      <EuiCompressedFieldNumber
                        name="detectionInterval"
                        id="detectionInterval"
                        placeholder="Interval"
                        data-test-subj="detectionInterval"
                        min={1}
                        style={{ width: numberFieldWidth }}
                        {...field}
                        onChange={handleIntervalChange}
                      />
                    </EuiFlexItem>
                    <EuiFlexItem>
                      <EuiText>
                        <p className="minutes">minutes</p>
                      </EuiText>
                    </EuiFlexItem>
                  </EuiFlexGroup>
                </FormattedFormRow>
              </EuiFlexItem>
            </EuiFlexGroup>
          );
        }}
      </Field>

      <Field name="frequency" validate={validateFrequency}>
        {({ field, form }: FieldProps) => (
          <EuiFlexGroup style={{ marginTop: '16px' }}>
            <EuiFlexItem style={{ maxWidth: '70%' }}>
              <FormattedFormRow
                fullWidth
                title="Frequency"
                hint={[
                  `Frequency sets how often the detector queries and scores data—i.e., how often alerts may fire. It must be a multiple of the interval and defaults to the same value. Choosing a longer frequency than the interval can improve efficiency by batching short intervals when ultra-fast alerting isn’t required (e.g., monitoring daily counts of anomalies across detectors) and is also useful for infrequent or batch log ingestion where data arrives irregularly or in bulk (e.g., a single daily ingestion spanning the entire day).`,
                ]}
                hintLink={`${AD_DOCS_LINK}`}
                isInvalid={isInvalid(field.name, form)}
                error={getError(field.name, form)}
              >
                <EuiFlexGroup gutterSize="s" alignItems="center">
                  <EuiFlexItem grow={false}>
                    <EuiCompressedFieldNumber
                      name="frequency"
                      id="frequency"
                      placeholder="Frequency"
                      data-test-subj="frequency"
                      min={1}
                      style={{ width: numberFieldWidth }}
                      {...field}
                    />
                  </EuiFlexItem>
                  <EuiFlexItem>
                    <EuiText>
                      <p className="minutes">minutes</p>
                    </EuiText>
                  </EuiFlexItem>
                </EuiFlexGroup>
              </FormattedFormRow>
            </EuiFlexItem>
          </EuiFlexGroup>
        )}
      </Field>

      <Field name="windowDelay" validate={validateEmptyOrNonNegativeInteger}>
        {({ field, form }: FieldProps) => (
          <FormattedFormRow
            fullWidth
            title="Window delay"
            hint="Specify a window of delay for a detector to fetch data, if you need to account for extra processing time."
            hintLink={`${AD_DOCS_LINK}`}
            isInvalid={isInvalid(field.name, form)}
            error={getError(field.name, form)}
            style={{ marginTop: '16px' }}
          >
            <EuiFlexGroup gutterSize="s" alignItems="center">
              <EuiFlexItem grow={false}>
                <EuiCompressedFieldNumber
                  name="windowDelay"
                  id="windowDelay"
                  placeholder="Window delay"
                  data-test-subj="windowDelay"
                  style={{ width: numberFieldWidth }}
                  {...field}
                />
              </EuiFlexItem>
              <EuiFlexItem>
                <EuiText>
                  <p className="minutes">minutes</p>
                </EuiText>
              </EuiFlexItem>
            </EuiFlexGroup>
          </FormattedFormRow>
        )}
      </Field>

      <Field name="history" validate={validateEmptyOrPositiveInteger}>
        {({ field, form }: FieldProps) => (
          <FormattedFormRow
            fullWidth
            title="History"
            hint="How far back the model looks for training data. This determines the amount of historical data used to train the anomaly detection model. Minimum history is 40 intervals, maximum is 10,000 intervals, and the default is 40."
            hintLink={`${AD_DOCS_LINK}`}
            isInvalid={isInvalid(field.name, form)}
            error={getError(field.name, form)}
            style={{ marginTop: '16px' }}
          >
            <EuiFlexGroup gutterSize="s" alignItems="center">
              <EuiFlexItem grow={false}>
                <EuiCompressedFieldNumber
                  name="history"
                  id="history"
                  placeholder="History"
                  data-test-subj="history"
                  min={40}
                  max={10000}
                  style={{ width: numberFieldWidth }}
                  {...field}
                />
              </EuiFlexItem>
              <EuiFlexItem>
                <EuiText>
                  <p className="minutes">intervals</p>
                </EuiText>
              </EuiFlexItem>
            </EuiFlexGroup>
          </FormattedFormRow>
        )}
      </Field>
    </Fragment>
  );
};
