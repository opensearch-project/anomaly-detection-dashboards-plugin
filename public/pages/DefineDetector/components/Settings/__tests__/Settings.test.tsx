/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

/*
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { Settings } from '../Settings';

import { Formik } from 'formik';

describe('<Settings /> spec', () => {
  test('renders the component', () => {
    const { container } = render(
      <Formik initialValues={{ detectorName: '' }} onSubmit={jest.fn()}>
        {() => (
          <div>
            <Settings />
          </div>
        )}
      </Formik>
    );
    expect(container.firstChild).toMatchSnapshot();
  });
  test('shows error for empty interval when toggling focus/blur', async () => {
    const { queryByText, findByText, getByPlaceholderText } = render(
      <Formik initialValues={{ name: '' }} onSubmit={jest.fn()}>
        {() => (
          <div>
            <Settings />
          </div>
        )}
      </Formik>
    );
    expect(queryByText('Required')).toBeNull();
    fireEvent.focus(getByPlaceholderText('Detector interval'));
    fireEvent.blur(getByPlaceholderText('Detector interval'));
    expect(findByText('Required')).not.toBeNull();
  });
  test('shows error for invalid interval when toggling focus/blur', async () => {
    const { queryByText, findByText, getByPlaceholderText } = render(
      <Formik initialValues={{ name: '' }} onSubmit={jest.fn()}>
        {() => (
          <div>
            <Settings />
          </div>
        )}
      </Formik>
    );
    expect(queryByText('Required')).toBeNull();
    userEvent.type(getByPlaceholderText('Detector interval'), '-1');
    fireEvent.blur(getByPlaceholderText('Detector interval'));
    expect(findByText('Must be a positive integer')).not.toBeNull();
  });
  test('shows error for interval of 0 when toggling focus/blur', async () => {
    const { queryByText, findByText, getByPlaceholderText } = render(
      <Formik initialValues={{ name: '' }} onSubmit={jest.fn()}>
        {() => (
          <div>
            <Settings />
          </div>
        )}
      </Formik>
    );
    expect(queryByText('Required')).toBeNull();
    userEvent.type(getByPlaceholderText('Detector interval'), '0');
    fireEvent.blur(getByPlaceholderText('Detector interval'));
    expect(findByText('Must be a positive integer')).not.toBeNull();
  });
  test('shows error for empty window delay when toggling focus/blur', async () => {
    const { queryByText, findByText, getByPlaceholderText } = render(
      <Formik initialValues={{ name: '' }} onSubmit={jest.fn()}>
        {() => (
          <div>
            <Settings />
          </div>
        )}
      </Formik>
    );
    expect(queryByText('Required')).toBeNull();
    fireEvent.focus(getByPlaceholderText('Window delay'));
    fireEvent.blur(getByPlaceholderText('Window delay'));
    expect(findByText('Required')).not.toBeNull();
  });
  test('shows error for invalid window delay when toggling focus/blur', async () => {
    const { queryByText, findByText, getByPlaceholderText } = render(
      <Formik initialValues={{ name: '' }} onSubmit={jest.fn()}>
        {() => (
          <div>
            <Settings />
          </div>
        )}
      </Formik>
    );
    expect(queryByText('Required')).toBeNull();
    userEvent.type(getByPlaceholderText('Window delay'), '-1');
    fireEvent.blur(getByPlaceholderText('Window delay'));
    expect(findByText('Must be a non-negative integer')).not.toBeNull();
  });
});
