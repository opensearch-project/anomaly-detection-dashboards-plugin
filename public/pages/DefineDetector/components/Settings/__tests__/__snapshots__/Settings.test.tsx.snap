// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Settings /> spec renders the component 1`] = `
<div>
  <div
    class="euiPanel euiPanel--paddingMedium euiPanel--borderRadiusMedium euiPanel--plain euiPanel--hasShadow"
    style="padding: 20px;"
  >
    <div
      class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
      style="padding: 0px;"
    >
      <div
        class="euiFlexItem"
      >
        <h3
          class="euiTitle euiTitle--small"
          data-test-subj="contentPanelTitle"
        >
          Operation settings
        </h3>
        <div
          class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--directionRow euiFlexGroup--responsive"
        >
          <div
            class="euiFlexItem content-panel-subTitle"
            style="line-height: normal; max-width: 75%;"
          />
        </div>
      </div>
      <div
        class="euiFlexItem euiFlexItem--flexGrowZero"
      >
        <div
          class="euiFlexGroup euiFlexGroup--gutterMedium euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
        >
          <div
            class="euiFlexItem"
          />
        </div>
      </div>
    </div>
    <div>
      <hr
        class="euiHorizontalRule euiHorizontalRule--full euiHorizontalRule--marginSmall"
      />
      <div
        style="padding: 10px 0px;"
      >
        <div
          class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--directionRow euiFlexGroup--responsive"
        >
          <div
            class="euiFlexItem"
            style="max-width: 70%;"
          >
            <div
              class="euiFormRow euiFormRow--fullWidth"
              hint="Define how often the detector collects data to generate anomalies. The shorter the interval is, the more real time the detector results will be, and the more computing resources the detector will need."
              hintlink="https://opensearch.org/docs/monitoring-plugins/ad"
              id="random_html_id-row"
              title="Detector interval"
            >
              <div
                class="euiFormRow__labelWrapper"
              >
                <label
                  aria-invalid="false"
                  class="euiFormLabel euiFormRow__label"
                  for="random_html_id"
                >
                  <div
                    style="line-height: 8px;"
                  >
                    <p>
                      Detector interval
                    </p>
                    <br />
                    <div
                      class="euiText euiText--medium sublabel"
                      style="max-width: 400px;"
                    >
                      Define how often the detector collects data to generate anomalies. The shorter the interval is, the more real time the detector results will be, and the more computing resources the detector will need.
                       
                      <a
                        class="euiLink euiLink--primary"
                        href="https://opensearch.org/docs/monitoring-plugins/ad"
                        rel="noopener noreferrer"
                        target="_blank"
                      >
                        Learn more
                        <svg
                          aria-hidden="true"
                          aria-label="External link"
                          class="euiIcon euiIcon--small euiIcon-isLoading euiLink__externalIcon"
                          focusable="false"
                          height="16"
                          role="img"
                          viewBox="0 0 16 16"
                          width="16"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <path
                            d="M5.277 10.088c.02.014.04.03.057.047.582.55 1.134.812 1.666.812.586 0 1.84-.293 3.713-.88L9 6.212V2H7v4.212l-1.723 3.876Zm-.438.987L3.539 14h8.922l-1.32-2.969C9.096 11.677 7.733 12 7 12c-.74 0-1.463-.315-2.161-.925ZM6 2H5V1h6v1h-1v4l3.375 7.594A1 1 0 0 1 12.461 15H3.54a1 1 0 0 1-.914-1.406L6 6V2Z"
                          />
                        </svg>
                        <span
                          class="euiScreenReaderOnly"
                        >
                          (opens in a new tab or window)
                        </span>
                      </a>
                    </div>
                  </div>
                </label>
              </div>
              <div
                class="euiFormRow__fieldWrapper"
              >
                <div
                  class="euiFlexGroup euiFlexGroup--gutterSmall euiFlexGroup--alignItemsCenter euiFlexGroup--directionRow euiFlexGroup--responsive"
                  id="random_html_id"
                >
                  <div
                    class="euiFlexItem euiFlexItem--flexGrowZero"
                  >
                    <div
                      class="euiFormControlLayout"
                    >
                      <div
                        class="euiFormControlLayout__childrenWrapper"
                      >
                        <input
                          class="euiFieldNumber"
                          data-test-subj="detectionInterval"
                          id="detectionInterval"
                          min="1"
                          name="interval"
                          placeholder="Detector interval"
                          type="number"
                          value=""
                        />
                      </div>
                    </div>
                  </div>
                  <div
                    class="euiFlexItem"
                  >
                    <div
                      class="euiText euiText--medium"
                    >
                      <p
                        class="minutes"
                      >
                        minutes
                      </p>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="euiFormRow euiFormRow--fullWidth"
          hint="Specify a window of delay for a detector to fetch data, if you need to account for extra processing time."
          hintlink="https://opensearch.org/docs/monitoring-plugins/ad"
          id="random_html_id-row"
          style="margin-top: 16px;"
          title="Window delay"
        >
          <div
            class="euiFormRow__labelWrapper"
          >
            <label
              aria-invalid="false"
              class="euiFormLabel euiFormRow__label"
              for="random_html_id"
            >
              <div
                style="line-height: 8px;"
              >
                <p>
                  Window delay
                </p>
                <br />
                <div
                  class="euiText euiText--medium sublabel"
                  style="max-width: 400px;"
                >
                  Specify a window of delay for a detector to fetch data, if you need to account for extra processing time.
                   
                  <a
                    class="euiLink euiLink--primary"
                    href="https://opensearch.org/docs/monitoring-plugins/ad"
                    rel="noopener noreferrer"
                    target="_blank"
                  >
                    Learn more
                    <svg
                      aria-hidden="true"
                      aria-label="External link"
                      class="euiIcon euiIcon--small euiIcon-isLoading euiLink__externalIcon"
                      focusable="false"
                      height="16"
                      role="img"
                      viewBox="0 0 16 16"
                      width="16"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M5.277 10.088c.02.014.04.03.057.047.582.55 1.134.812 1.666.812.586 0 1.84-.293 3.713-.88L9 6.212V2H7v4.212l-1.723 3.876Zm-.438.987L3.539 14h8.922l-1.32-2.969C9.096 11.677 7.733 12 7 12c-.74 0-1.463-.315-2.161-.925ZM6 2H5V1h6v1h-1v4l3.375 7.594A1 1 0 0 1 12.461 15H3.54a1 1 0 0 1-.914-1.406L6 6V2Z"
                      />
                    </svg>
                    <span
                      class="euiScreenReaderOnly"
                    >
                      (opens in a new tab or window)
                    </span>
                  </a>
                </div>
              </div>
            </label>
          </div>
          <div
            class="euiFormRow__fieldWrapper"
          >
            <div
              class="euiFlexGroup euiFlexGroup--gutterSmall euiFlexGroup--alignItemsCenter euiFlexGroup--directionRow euiFlexGroup--responsive"
              id="random_html_id"
            >
              <div
                class="euiFlexItem euiFlexItem--flexGrowZero"
              >
                <div
                  class="euiFormControlLayout"
                >
                  <div
                    class="euiFormControlLayout__childrenWrapper"
                  >
                    <input
                      class="euiFieldNumber"
                      data-test-subj="windowDelay"
                      id="windowDelay"
                      name="windowDelay"
                      placeholder="Window delay"
                      type="number"
                      value=""
                    />
                  </div>
                </div>
              </div>
              <div
                class="euiFlexItem"
              >
                <div
                  class="euiText euiText--medium"
                >
                  <p
                    class="minutes"
                  >
                    minutes
                  </p>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
